Systemd unit creation stage

say-paul opened the pull request:
This stage will let user to create  systemd service in '/usr/lib/systemd/system'
Stage accepts:
'Filename' - string , must end with '.service'
'Unit' - section
'Service' - section
'Install'- section

thozza reviewed (dismissed):
Thanks for your contribution.

Unfortunately, renaming an existing osbuild stage and replacing it with an incompatible stage is not really an option.

This is against osbuild principles (https://github.com/osbuild/osbuild/pull/1568). Specifically, old manifests must not be broken with new osbuild.

The only option is to:

1. Extend the `org.osbuild.systemd.unit` stage to handle both situations, the creation of units as well as dropins, in a backward compatible way. I think that this can be done in a reasonable way by making the `dropin` stage option optional.
2. Introduce a completely new stage (e.g. `org.osbuild.systemd.unit.create`), which will handle only the creation of new unit files.

say-paul commented:
> Introduce a completely new stage (e.g. org.osbuild.systemd.unit.create), which will handle only the creation of new unit files.

my vote is for this, as overriding a drop-in will make things bit complicated to maintain.

say-paul commented on stages/org.osbuild.systemd.unit.create:
```diff
@@ -0,0 +1,102 @@
+  "config": {
+    "additionalProperties": false,
+    "type": "object",
+    "description": "Drop-in configuration for a '.service' unit.",
```
remove Drop-in

mvo5 commented:
I'm sorry that I miss some context here. What problem/use-case will this stage solve? I want to help but I'm slightly worried about how open ended this stage is, if we allow arbitrary systemd units with arbitrary dependencies and scripts in them can easily break the boot of a system and it's hard to audit and debug and hard to know beforehand what will work and what won't. Arguably this is true for many of our stages so not a reason as such to not do it. But it seems in general we try to be more declarative than imperative so maybe there is something less open we could do. OTOH we already have the drop-in stage which allows similar things (slightly less powerful but still pretty powerful). Sorry for the rambling, understanding this a bit better would help me :)

say-paul commented:
> I'm sorry that I miss some context here. What problem/use-case will this stage solve?

This is for cutting out a stage to put a systemd service that is  required for : https://github.com/osbuild/images/pull/399. 

> I  want to help but I'm slightly worried about how open ended this stage is

I have kept it open as systemd options are vast an it will be very long schema is want to put everything.
I am open for suggestions though if we want to limit to only things that is required for now to add some sort of `systemd-analyze verify` .

ondrejbudai commented on stages/org.osbuild.systemd.unit.create:
```diff
@@ -0,0 +1,102 @@
+            key_value_pair = str(opt).split("=", 1)  # maxsplit set to split 1st occurance of '='
+            if len(key_value_pair) == 1:
+                return 1
+            config.set(section, key_value_pair[0], key_value_pair[1])
```
Just a random thought: systemd allows specifying assignments multiple times, but configparser doesn't support it. If we want to expose this option, we might need to use something else than configparser. See https://bugs.python.org/issue31528 and

```
[Service]  
Type=oneshot
ExecStart=mkfs.ext4 /dev/sdb
ExecStart=mkdir /mnt/data
ExecStart=mount /dev/sdb /mnt/data
ExecStart=bash -c "echo '/dev/sdb /mnt/data ext4 defaults 0 2' >>/etc/fstab"
ExecStartPost=touch /var/lib/prepare-data-disk-first-boot
```

say-paul replied:
Nice catch!

say-paul replied:
Quick solution was to allow_now_value in config parser and put the whole string as option without value.
I think we wont have duplicate systemd option:value combo under a section.

mvo5 commented:
> > I'm sorry that I miss some context here. What problem/use-case will this stage solve?
> 
> This is for cutting out a stage to put a systemd service that is required for : [osbuild/images#399](https://github.com/osbuild/images/pull/399).

Thank you! This is very useful background for me.
 
> > I  want to help but I'm slightly worried about how open ended this stage is
> 
> I have kept it open as systemd options are vast an it will be very long schema is want to put everything. I am open for suggestions though if we want to limit to only things that is required for now to add some sort of `systemd-analyze verify` .

Thank you again - I realize I was too strict in my initial comment, sorry for that.

thozza commented:
> > I'm sorry that I miss some context here. What problem/use-case will this stage solve?
> 
> This is for cutting out a stage to put a systemd service that is required for : [osbuild/images#399](https://github.com/osbuild/images/pull/399).
> 
> > I  want to help but I'm slightly worried about how open ended this stage is
> 
> I have kept it open as systemd options are vast an it will be very long schema is want to put everything. I am open for suggestions though if we want to limit to only things that is required for now to add some sort of `systemd-analyze verify` .

TBH, I'm not a big fan of this current approach. The problem is that one can create invalid unit file with this stage, which we should prevent. We tend to list all options explicitly in the stage options schema. While I understand that the options are vast for systemd units, you can start only with the subset that you need for your needs.

Also, the schema needs to be more restrictive, e.g. it must require that the minimal required sections are specified, that they are not empty (if not valid), etc..

say-paul commented:
> TBH, I'm not a big fan of this current approach. The problem is that one can create invalid unit file with this stage, which we should prevent. We tend to list all options explicitly in the stage options schema. While I understand that the options are vast for systemd units, you can start only with the subset that you need for your feature.

I am thinking of adding a `systemd-analyze verify <file>`, to safe guard against any malformed systemd unit. I does offer a level of sanity-check.

@thozza @mvo5  thoughts?

say-paul commented:
Something like this:

```
org.osbuild.systemd.unit.create: e8f7850b43b4acd74ea2d79de7872902588e98cd23474856cc8003775be65ba7 {
  "filename": "create-directory.service",
  "config": {
    "Unit": [
      "Description=Create directory"
    ],
    "Service": [
      "Type=onesho",
      "RemainAfterExit=yes",
      "ExecStart=mkdir -p /var/myfiles",
      "ExecStart=mkdir /mnt/data"
    ],
    "Install": [
      "WantedBy=local-fs.target"
    ]
  }
}
Failed to open file "/sys/fs/selinux/checkreqprot": Read-only file system
/run/osbuild/tree/usr/lib/systemd/system/create-directory.service:5: Failed to parse service type, ignoring: onesho
create-directory.service: Service has more than one ExecStart= setting, which is only allowed for Type=oneshot services. Refusing.
Unit create-directory.service has a bad unit file setting.
```

achilleas-k commented:
> Something like this:
> 
> ```
> org.osbuild.systemd.unit.create: e8f7850b43b4acd74ea2d79de7872902588e98cd23474856cc8003775be65ba7 {
>   "filename": "create-directory.service",
>   "config": {
>     "Unit": [
>       "Description=Create directory"
>     ],
>     "Service": [
>       "Type=onesho",
>       "RemainAfterExit=yes",
>       "ExecStart=mkdir -p /var/myfiles",
>       "ExecStart=mkdir /mnt/data"
>     ],
>     "Install": [
>       "WantedBy=local-fs.target"
>     ]
>   }
> }
> Failed to open file "/sys/fs/selinux/checkreqprot": Read-only file system
> /run/osbuild/tree/usr/lib/systemd/system/create-directory.service:5: Failed to parse service type, ignoring: onesho
> create-directory.service: Service has more than one ExecStart= setting, which is only allowed for Type=oneshot services. Refusing.
> Unit create-directory.service has a bad unit file setting.
> ```

The `verify` isn't a bad idea, but in this particular example, I'd expect osbuild to fail to verify to begin with.  The `Type` should be an enum of `simple, exec, forking, oneshot, dbus, notify, notify-reload, or idle` (systemd.service(5)).

Since we can't hope to cover all verification in the stage, let alone the schema, I think verifying with `systemd-analyze` would be good, but obvious things, like an invalid value for a key that has a very explicit set of allowed values, would be caught immediately before osbuild even runs.

achilleas-k reviewed (dismissed):
> TBH, I'm not a big fan of this current approach. The problem is that one can create invalid unit file with this stage, which we should prevent. We tend to list all options explicitly in the stage options schema. While I understand that the options are vast for systemd units, you can start only with the subset that you need for your feature.

Sorry I think my previous message was a bit misleading.  I was in favour of adding the verification step at the end but not at the cost of what @thozza said here.  I really don't like the key-value pairs being arbitrary strings that just get written to the file as-is.  I'd like this to be as explicit as possible.  And since we don't need that many options to begin with, I think it's okay to start small and keep adding as necessary.  So for now, the schema can reflect the unit file structure:
```json
"additionalProperties": false,
"required": ["filename","config"],
"properties": {
  "filename": {
    "type": "string",
    "pattern": "^[\\w:.\\\\-]+[@]{0,1}[\\w:.\\\\-]*\\.service$"
  },
  "config": {
    "additionalProperties": false,
    "type": "object",
    "description": "Configuration for a '.service' unit.",
    "properties": {
      "Type": {
        "type": "string",
        "enum": ["simple", "exec", "forking", "oneshot", "dbus", "notify", "notify-reload", "idle"]
      },
      "Unit": {
        "additionalProperties": false,
        "type": "object",
        "description": "'Unit' configuration section of a unit file.",
        "properties": {
          "Description": {
            "type": "string"
          },
          "Wants": {
            "type": "array",
            "items": {
              "type": "string"
            }
          }
        }
      },
      "Service": {
        "additionalProperties": false,
        "type": "object",
        "description": "'Service' configuration section of a unit file.",
        "properties": {
          "RemainAfterExit": {
            "type": "boolean"
          },
          "ExecStart": {
            "type": "array",
            "items": {
              "type": "string"
            }
          }
        }
      },
      "Install": {
        "additionalProperties": false,
        "type": "object",
        "description": "'Install' configuration section of a unit file.",
        "properties": {
          "WantedBy": {
            "type": "array",
            "items": {
              "type": "string"
            }
          }
        }
      }
    }
  }
}
```

achilleas-k reviewed (dismissed):
Thanks for all the work!  I really like how this turned out now.  LGTM!

mvo5 commented on stages/org.osbuild.systemd.unit.create:
```diff
@@ -0,0 +1,181 @@
+    with open(f"{systemd_dir}/{filename}", "w", encoding="utf8") as f:
+        config.write(f, space_around_delimiters=False)
+
+    ret = subprocess.run(["systemd-analyze", "verify", f"{systemd_dir}/{filename}"], check=True)
```
The `check=True` and `return ret.returncode` is a bit redundant here as any non-zero return code will raise an error.

mvo5 reviewed (commented):
Thanks for this PR! This looks very nice now. I added some small questions/suggestions inline. I would love to see a test as well, I attached a small commit with what I have in mind, feel free to use or modify as you see fit!

[0001-stages-org.osbuild.systemd.unit.create-add-small-uni.patch.txt](https://github.com/osbuild/osbuild/files/14283353/0001-stages-org.osbuild.systemd.unit.create-add-small-uni.patch.txt)

mvo5 commented:
Just to be clear, test can be a followup, I don't want to block this PR because of this (and same goes for my questions, nothing here seems like a blocker).

say-paul commented on stages/org.osbuild.systemd.unit.create:
```diff
@@ -0,0 +1,181 @@
+    "config": {
+        "additionalProperties": false,
+        "type": "object",
+        "description": "Configuration for a '.service' unit.",
```
that's a good point, we can add `"required": ["Unit", "Service", "Install"]` ensure that.

say-paul commented on stages/org.osbuild.systemd.unit.create:
```diff
@@ -0,0 +1,181 @@
+    filename = options["filename"]
+    cfg = options["config"]
+    # ensure the service name does not exceed maximum filename length
+    if len(filename) > 255:
```
I guess its personal preference, I feel assigning variable data at top makes it easier to arrange scripts. I am not aware of any best practices though, as its a trivial change I can put it along with the other changes.

mvo5 replied:
Thanks for changing - just for the record it's really totally fine either way(this is why I called it a "super nitpick") :) - I would have not insisted on this (but I appreciate that you changed it!)

say-paul commented:
> Thanks for this PR! This looks very nice now. I added some small questions/suggestions inline. I would love to see a test as well, I attached a small commit with what I have in mind, feel free to use or modify as you see fit!
> 
> [0001-stages-org.osbuild.systemd.unit.create-add-small-uni.patch.txt](https://github.com/osbuild/osbuild/files/14283353/0001-stages-org.osbuild.systemd.unit.create-add-small-uni.patch.txt)

added issue to track this: #1591

mvo5 reviewed (approved):
Thank you, this looks fine now, once it's in I will push a small followup with the test

achilleas-k reviewed (dismissed):
LGTM still.  Let's get this in!

achilleas-k commented:
I hope we can get the F40 tests sorted quickly, otherwise it might we worth skipping them for this PR and enabling them after.  Not sure what's best here.

achilleas-k commented:
Opened https://github.com/osbuild/osbuild/pull/1595 to get the F40 repos sorted (hopefully).  Cherry picking the same commit here to get this going.

croissanne commented on Schutzfile:
```diff
@@ -179,14 +179,14 @@
           {
             "title": "fedora",
             "name": "fedora",
-            "baseurl": "https://rpmrepo.osbuild.org/v2/mirror/public/f40/f40-x86_64-rawhide-20240101"
```
ugh we should just have a 'rawhide' branch until f40 comes out and then do f40. Instead of this -rawhide and -branched.

achilleas-k replied:
I just copied what we did last time in rpmrepo.
